tools/agentlint/internal/rivertags/rivertags.go

```diff
@@ -0,0 +1,248 @@
+			if len(matches) == 0 {
+				continue
+			} else if len(matches) > 1 {
+				p.Report(analysis.Diagnostic{
+					Pos:      field.Pos(),
+					Category: "rivertags",
+					Message:  "field should not have more than one river tag",
+				})
+			}
+
+			// Before checking the tag, do general validations first.
+			if field.Anonymous() {
+				p.Report(analysis.Diagnostic{
+					Pos:      field.Pos(),
+					Category: "rivertags",
+					Message:  "river tags may not be given to anonymous fields",
+				})
+			}
+			if !field.Exported() {
+				p.Report(analysis.Diagnostic{
+					Pos:      field.Pos(),
+					Category: "rivertags",
+					Message:  "river tags may only be given to exported fields",
+				})
+			}
+			if len(nodeField.Names) > 1 {
+				// Report "a, b, c int `river:"name,attr"`" as invalid usage.
+				p.Report(analysis.Diagnostic{
+					Pos:      field.Pos(),
+					Category: "rivertags",
+					Message:  "river tags should not be inserted on field names separated by commas",
+				})
+			}
+
+			for _, match := range matches {
+				diagnostics := lintRiverTag(field, match[1])
+				for _, diag := range diagnostics {
+					p.Report(analysis.Diagnostic{
+						Pos:      field.Pos(),
+						Category: "rivertags",
+						Message:  diag,
+					})
+				}
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func getStructs(ti *types.Info) []*structInfo {
+	var res []*structInfo
+
+	for ty, def := range ti.Defs {
+		def, ok := def.(*types.TypeName)
+		if !ok {
+			continue
+		}
+
+		structTy, ok := def.Type().Underlying().(*types.Struct)
+		if !ok {
+			continue
+		}
+
+		switch node := ty.Obj.Decl.(*ast.TypeSpec).Type.(type) {
+		case *ast.StructType:
+			res = append(res, &structInfo{
+				Node: node,
+				Type: structTy,
+			})
+		default:
+		}
+	}
+
+	return res
+}
+
+func lookupField(node *ast.StructType, index int) *ast.Field {
+	startIndex := 0
+
+	for _, f := range node.Fields.List {
+		length := len(f.Names)
+		if length == 0 { // Embedded field
+			length = 1
+		}
+
+		endIndex := startIndex + length
+		if index >= startIndex && index < endIndex {
+			return f
+		}
+
+		startIndex += length
+	}
+
+	panic(fmt.Sprintf("index %d out of range %d", index, node.Fields.NumFields()))
+}
+
+type structInfo struct {
+	Node *ast.StructType
+	Type *types.Struct
+}
+
+func lintRiverTag(ty *types.Var, tag string) (diagnostics []string) {
+	if tag == "" {
+		diagnostics = append(diagnostics, "river tag should not be empty")
+		return
+	}
+
+	parts := strings.SplitN(tag, ",", 2)
+	if len(parts) != 2 {
+		diagnostics = append(diagnostics, "river tag is missing options")
+		return
+	}
+
+	var (
+		name    = parts[0]
+		options = parts[1]
+
+		nameParts = splitName(name)
+	)
+
+	switch options {
+	case "attr", "attr,optional":
+		if len(nameParts) == 0 {
+			diagnostics = append(diagnostics, "attr field must have a name")
+		} else if len(nameParts) > 1 {
+			diagnostics = append(diagnostics, "attr field names must not contain `.`")
+		}
+		for _, name := range nameParts {
```

Comment:
Yeah, I'm iterating over it anyway just in case they incorrectly put more than one, it's still useful (imo) to show the various errors across all the name fragments..